### PR TITLE
kind integrated test workflow MVP

### PR DESCRIPTION
**Pull request by BenTheElder**
implements #986

### CODE REVIEW

File: hack/ci/e2e-k8s.sh [hunk 1]
```diff
@@ -22,167 +22,9 @@ set -o errexit -o nounset -o xtrace
...
-  # create the cluster and run tests
-  create_cluster && run_tests
-}
-
-main
+# BUILD_TYPE: bazel or docker
+#
+export KIND_BUILD_TYPE="${BUILD_TYPE:-bazel}"
+export KIND_TEST_FOCUS_OVERRIDE="${FOCUS:-}"
+export KIND_TEST_SKIP_OVERRIDE="${SKIP:-}"
```
Comment (BenTheElder) [inline hack/ci/e2e-k8s.sh]: in the future we will stop doing this and just control the lists in code centrally (IE `kind test presubmit` will have a fixed list, with the override being meant for local prototyping hacks, not CI)

File: hack/ci/e2e-k8s.sh [hunk 2]
```diff
@@ -22,167 +22,9 @@ set -o errexit -o nounset -o xtrace
...
+kind test presubmit --ip-family="${IP_FAMILY:-ipv4}"
```
Comment (BenTheElder) [inline hack/ci/e2e-k8s.sh]: intend to minimize knobs that need touching heavily to avoid the typical confusion we have with our test rigs, however ip-family is one axis that we definitely test along in kubernetes presubmit and can't really handle well otherwise.

when we migrate this into prow config it will be simpler like:
`kind test presubmit`
`kind test presubmit --ip-family=ipv6`
w/o the env stuff.

Comment (aojea) [reply hack/ci/e2e-k8s.sh]: just that you mention this topic, want to share something I realized this week.
With dual-stack is it possible to have 2 kind of clusters:
- dualstack ipv4 (default) - ipv6 (sec)
- dualstack ipv6 (default) - ipv4 (sec)

maybe something we should think for the future, the default is giving by the order of the cluster-cidr configured.

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+			}
+			return errors.New("aborting due to invalid configuration")
+		}
+		return errors.Wrap(err, "failed to create cluster")
+	}
+
+	// In Kubernetes CI we don't have real IPv6 connectivity so we need to
+	// employ some work arounds when testing IPv6 clusters
+	// We do this all environments for consistency.
+	if config.Networking.IPFamily == v1alpha4.IPv6Family {
```
Comment (aojea) [inline pkg/cmd/kind/test/internal/testcmd/testcmd.go]: `config.Networking.IPFamily == v1alpha4.IPv6Family || config.Networking.IPFamily == v1alpha4.DualStackFamily`

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+		}
+	}
+
+	// run tests
+	skip := details.Skip
+	if details.Parallel {
+		skip = strings.Join([]string{`\[Serial\]`, skip}, "|")
+	}
+	cmd := exec.Command(
+		"./hack/ginkgo-e2e.sh",
```
Comment (aojea) [inline pkg/cmd/kind/test/internal/testcmd/testcmd.go]: do we need the  `./hack/ginkgo-e2e.sh` bash script?
a quick look to the script seems that we can avoid most of the checks and run directly the `e2e.test`binary,

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+	if err != nil {
+		return errors.Wrap(err, "failed to get control-plane node for coreDNS IPv6 CI fixes")
+	}
+	if len(nodes) < 1 {
+		return errors.New("failed to locate a kind node for coreDNS IPv6 CI fixes")
+	}
+	node := nodes[0]
+
+	// "fix" coreDNS configmap
+	// TODO: refactor this to not be a shell script.
```
Comment (aojea) [inline pkg/cmd/kind/test/internal/testcmd/testcmd.go]: +1000
This can also be a good /help wanted issue

File: pkg/cmd/kind/test/test.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+	"sigs.k8s.io/kind/pkg/log"
+)
+
+// NewCommand returns a new cobra.Command for test
+func NewCommand(logger log.Logger, streams cmd.IOStreams) *cobra.Command {
+	cmd := &cobra.Command{
+		Args: cobra.NoArgs,
+		// TODO(bentheelder): more detailed usage
+		Use:   "test",
+		Short: "Tests one of [presubmit, conformance]",
```
Comment (aojea) [inline pkg/cmd/kind/test/test.go]: how can I run my custom tests? this is specially useful to troubleshoot e2e tests or to create new ones, where you only want to run 1 or a small number of tests.
can we add one subcommand that allows to specify my custom FOCUS and SKIP fields?

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+	skip := details.Skip
+	if details.Parallel {
+		skip = strings.Join([]string{`\[Serial\]`, skip}, "|")
+	}
+	cmd := exec.Command(
+		"./hack/ginkgo-e2e.sh",
+		// don't use the horrid e2e.test """provider""" concept
+		"--provider=skeleton",
+		// need to inform tests of worker count
+		"--num-nodes=", strconv.Itoa(numNodesForE2E),
```
Comment (aojea) [inline pkg/cmd/kind/test/internal/testcmd/testcmd.go]: the jobs are failing with
> I0118 01:25:13.095] [1] invalid value "" for flag -num-nodes: strconv.ParseInt: parsing "": invalid syntax
I0118 01:25:13.095] [1] Usage of /go/src/k8s.io/kubernetes/bazel-bin/test/e2e/e2e.test:
🤷‍♂

Comment (aojea) [reply pkg/cmd/kind/test/internal/testcmd/testcmd.go]: ok, there is the problem, is adding an space
```
W0118 01:25:14.472] Call stack:
W0118 01:25:14.472]   1: ./hack/ginkgo-e2e.sh:143 main(...)
W0118 01:25:14.472] Exiting with status 1
W0118 01:25:22.159] ERROR: command "./hack/ginkgo-e2e.sh --provider=skeleton --num-nodes= 2 --report-dir /workspace/_artifacts --disable-log-dump=true --ginkgo.focus '' --ginkgo.skip '\[Serial\]|'" failed with error: exit status 1
```

Comment (BenTheElder) [reply pkg/cmd/kind/test/internal/testcmd/testcmd.go]: no, the space is a rendering of passing this to the shell.
the problem is that we're passing it as two arguments, but the first contains `=`

marked the PR WIP, I haven't even run this yet. by inspection it's mostly done but it's still a prototype. I'm going to rework things a bit before un-WIP-ing it.

**Comment by BenTheElder** [comment]
e2e is failing because kubectl isn't being found on the path when building with bazel + running ginkgo-e2e.sh. probably not actually fixed upstream :/
will settle on an option for this tomorrow.

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+	cleanup := func(calledFromSignalHandler bool) {
+		err2 := provider.CollectLogs(flags.Name, filepath.Join(flags.Artifacts, "logs"))
+		err3 := provider.Delete(flags.Name, flags.kubeconfigPath())
+		// don't touch err if we're being called from another goroutine
+		// on program early exit, or if err is already non-nil
+		if calledFromSignalHandler || err != nil {
+			return
+		} else if err2 != nil {
+			err = err2
+		} else {
```
Comment (aojea) [inline pkg/cmd/kind/test/internal/testcmd/testcmd.go]: this is only `if err3 != nil` or we alway assign err3 to err at this point?

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+		return errors.Wrap(err, "error creating build context")
+	}
+	if err := ctx.Build(); err != nil {
+		return errors.Wrap(err, "error building node image")
+	}
+
+	// build test binaries
+	switch buildType {
+	case "bazel":
+		if err := exec.InheritOutput(
```
Comment (aojea) [inline pkg/cmd/kind/test/internal/testcmd/testcmd.go]: this writes to the logger, right?

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"sigs.k8s.io/kind/pkg/apis/config/v1alpha4"
```
Comment (aojea) [inline pkg/cmd/kind/test/internal/testcmd/testcmd.go]: shouldn't we use the internal APIs?

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+
+func boolToYorN(y bool) string {
+	if y {
+		return "y"
+	}
+	return "n"
+}
+
+func enableIPv6Workarounds(provider *cluster.Provider, name string) error {
+	// get a control plane node to use run kubectl from
```
Comment (aojea) [inline pkg/cmd/kind/test/internal/testcmd/testcmd.go]: maybe we should check that ipv6 is enabled? or at least ipv6 forwarding?

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+	nodes, err := provider.ListInternalNodes(name)
+	if err != nil {
+		return errors.Wrap(err, "failed to get nodes for coreDNS IPv6 CI fixes")
+	}
+	nodes, err = nodeutils.ControlPlaneNodes(nodes)
```
Comment (aojea) [inline pkg/cmd/kind/test/internal/testcmd/testcmd.go]: nit, a different name to the assigned variable to distinguish from the previous one?

File: hack/ci/e2e-k8s.sh (hunk 2 above)
Comment (BenTheElder) [reply hack/ci/e2e-k8s.sh]: that isn't really related to this change though.

File: pkg/cmd/kind/test/test.go (hunk 6 above)
Comment (BenTheElder) [reply pkg/cmd/kind/test/test.go]: - see the e2e.sh for details on overriding these
- it's not intended to run _your_ tests. it's intended to run the standard flows in a way a non-power-user can trivially mimic by invoking the command(s). you as a power user can still write a little script with your flows.

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go (hunk 5 above)
Comment (BenTheElder) [reply pkg/cmd/kind/test/internal/testcmd/testcmd.go]: maybe? probably not. the replacements need to be equivalent. i'm not sure how many casual contributors know sed / shell and go string manipulation deeply 🤷‍♂

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go (hunk 12 above)
Comment (BenTheElder) [reply pkg/cmd/kind/test/internal/testcmd/testcmd.go]: that's anti-helpful here. we don't want to retain and accidentally use the unfiltered node list.
we could break this out into a helper method but it's just a few lines so that's not super helpful

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go (hunk 11 above)
Comment (BenTheElder) [reply pkg/cmd/kind/test/internal/testcmd/testcmd.go]: the caller is responsible for having ipv6 enabled docker. we're not going to muck with the host in `kind`

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go (hunk 4 above)
Comment (BenTheElder) [reply pkg/cmd/kind/test/internal/testcmd/testcmd.go]: that's orthogonal to this PR. but in any case I'm not convinced that this is useful right now, it will skew versus other test callers and this script also does helpful things like locating the e2e binary in the many possible build paths.

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go (hunk 3 above)
Comment (BenTheElder) [reply pkg/cmd/kind/test/internal/testcmd/testcmd.go]: we can't realy TODO constants that don't exist yet, dual stack my not be a family, it will probably be many families?
dual stack as an API is a TODO.

also, there is validation long before this code rune that errors if not one of ipv4 / ipv6 specifically to handle the fact that this command is not aware of any other families currently

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go (hunk 9 above)
Comment (BenTheElder) [reply pkg/cmd/kind/test/internal/testcmd/testcmd.go]: no. this wires it to stdout / stderr.

this respecting streams is a broad TODO, but it's not super important for cmd/ at the moment

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go (hunk 8 above)
Comment (BenTheElder) [reply pkg/cmd/kind/test/internal/testcmd/testcmd.go]: we don't care at this point, because if err3 is nil at this point so are err2 and err /shrug

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go (hunk 10 above)
Comment (BenTheElder) [reply pkg/cmd/kind/test/internal/testcmd/testcmd.go]: no. internal apis are for:
- translating between versions
- allowing runtime code implementing the API to handle a single version, possibly with extra (computed?) data that is not part of the public API.

the CLI is a client of the runtime.

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go (hunk 12 above)
Comment (aojea) [reply pkg/cmd/kind/test/internal/testcmd/testcmd.go]: nah, no problem at all, is just there are at least 1 or 2 places more doing this that way, that's why I added the nit  😄

File: hack/ci/e2e-k8s.sh (hunk 2 above)
Comment (aojea) [reply hack/ci/e2e-k8s.sh]: why I can't resolve the review comments to minimize them?
🤔

Comment (BenTheElder) [reply hack/ci/e2e-k8s.sh]: you don't have permissions. :github:

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go (hunk 3 above)
Comment (aojea) [reply pkg/cmd/kind/test/internal/testcmd/testcmd.go]: 👍

File: pkg/cmd/kind/test/internal/testcmd/testcmd.go (hunk 4 above)
Comment (aojea) [reply pkg/cmd/kind/test/internal/testcmd/testcmd.go]: 👍

File: pkg/cmd/kind/test/test.go (hunk 6 above)
Comment (aojea) [reply pkg/cmd/kind/test/test.go]: I think I've found my answer, just let me rephrase, how can I specify FOCUS and SKIP with this change?
I have to use the env variables `KIND_TEST_SKIP_OVERRIDE` and `KIND_TEST_FOCUS_OVERRIDE` right?

Comment (BenTheElder) [reply pkg/cmd/kind/test/test.go]: those are temporary for transitioning the prowjobs
you can run e2e.test / ginkgo e2e.sh yourself currently.

**Comment by aojea** [comment]
this is an awesome improvement, please disregard my comments, I don't know how to minimize them 🙃

**Comment by BenTheElder** [comment]
/test all

https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kind/1269/pull-kind-e2e-kubernetes/1220495992827154437

that is some weird test behavior

**Comment by BenTheElder** [comment]
[doing some experiments with e2e before moving forward, commits etc. will be cleaned up before hold / WIP are removed]

**Comment by aojea** [comment]
tests are failing because of 🤔 
`    exec: "kubectl": executable file not found in $PATH`

**Comment by BenTheElder** [comment]
poor connectivity and forgot to warm local deps right now so using CI to
debug

**Comment by BenTheElder** [comment]
trailing newline bug :P

**Comment by BenTheElder** [comment]
Just flakes. Will clean this up soon. The PATH hacks are lame and were not supposed to still be required, will write a more portable approach.

**Comment by aojea** [comment]
one small doc to understand how to use it or mostly for own consumption?

**Comment by BenTheElder** [comment]
Docs come when it's not an MVP :+)

**Comment by BenTheElder** [comment]
However it really is just `kind test presubmit` if you have the deps setup for building k8s. Other stuff is for CI, except the optional flags which have help usage ...

**Comment by BenTheElder** [comment]
/retest
Storage test flake in 1.14

**Comment by aojea** [comment]
Do you think that is possible to implement 2 executions of the `e2e.test`?
One using parallel without the serial tests and other only with the serial tests and aggregating  the results of both runs

**Comment by BenTheElder** [comment]
I don't think we should deviate from the rest of upstream currently.

**Comment by BenTheElder** [comment]
okaaay, let's see about getting this up to date again
